move and revise AWS constraints (#846)
aws/karpenter: website/content/en/pre-docs/reference/provisioner-crd.md

  # These requirements are combined with pod.spec.affinity.nodeAffinity rules.
  # Operators { In, NotIn } are supported to enable including or excluding values
  requirements:
    - key: "node.kubernetes.io/instance-type" # If not included, reference cloud provider
      operator: In
      values: ["m5.large", "m5.2xlarge"]
    - key: "topology.kubernetes.io/zone" # If not included, all zones are considered
      operator: In
      values: ["us-west-2a", "us-west-2b"]
    - key: "kubernetes.io/arch" # If not included, reference cloud provider
      operator: In
      values: ["arm64", "amd64"]
    - key: "kubernetes.io/os" # only linux is currently supported
      operator: In
      values: ["linux"]
    - key: "karpenter.sh/capacity-type" # If not included, reference the cloud provider
      operator: In
      values: ["spot", "on-demand"]
  # These fields vary per cloud provider, see your cloud provider specific documentation
  provider: {}
```

## spec.requirements

> Kubernetes defines the following [well known labels]([[link]]), and cloud providers (e.g., AWS) merely implement them. They are defined at the "spec.requirements" section of the provisioner CRD. 

Reviewer: delete "merely"

These well known labels may be specified at the provisioner level, or in a workload definition (e.g., nodeSelector on a pod.spec). If the two conflict, the provisioner controls. In other words, a podspec may not override the provisioner with respect to constraints. 

> For example, Instance type may also be specified using a nodeSelector in a workload definition. If the instance type requested at the workload (e.g., pod) level is not included in the provisioner list, karpenter will not create a node or schedule the pod. 

Reviewer: "For example, the instance type maybe specified using..."

```yaml
spec:
  requirements:
    - key: node.kubernetes.io/instance-type
      operator: In
      values: ["m5.large", "m5.2xlarge"]
```

*Override with workload manifest (e.g., pod)*

```yaml
spec:
  template:
    spec:
      nodeSelector:
        node.kubernetes.io/instance-type: m5.large
```

### Availability Zones

- key: `topology.kubernetes.io/zone`
- value example: `us-east-1c`

☁️ **AWS**

- value list: `aws ec2 describe-availability-zones --region <region-name>`

Karpenter can be configured to create nodes in a particular zone. Note that the Availability Zone us-east-1a for your AWS account might not have the same location as us-east-1a for another AWS account.

[Learn more about Availability Zone
IDs.](https://docs.aws.amazon.com/ram/latest/userguide/working-with-az-ids.html)

### Architecture

- key: `kubernetes.io/arch`
- values
  - `amd64` (default)
  - `arm64`

Karpenter supports `amd64` nodes, and `arm64` nodes.

### Operating System

- key: `kubernetes.io/os`
- values
  - `linux` (default)

At this time, Karpenter only supports Linux OS nodes.

### Capacity Type

- key: `karpenter.sh/capacity-type`

☁️ **AWS**

- values
  - `spot` (default)
  - `on-demand` 

> Karpenter supports specifying capacity type, which is analogous to EC2 usage classes (aka "market types").

Reviewer: I haven't heard of market types before. I think we usually refer to this as purchasing options.

> Karpenter defaults to spot instances. [Spot instances](https://aws.amazon.com/ec2/spot/) may be preempted, and should not be used for critical workloads that do not tolerate interruptions.

Reviewer: Karpenter defaults to spot for all instances it provision? If so, can be make this bold? I think this is important to call out.

Author: yes. i agree this is important to highlight.

  # These requirements are combined with pod.spec.affinity.nodeAffinity rules.
  # Operators { In, NotIn } are supported to enable including or excluding values
  requirements:

>     - key: "node.kubernetes.io/instance-type" # If not included, reference cloud provider

Reviewer 2: These comment changes are all very confusing to me now.

      operator: In
      values: ["m5.large", "m5.2xlarge"]
    - key: "topology.kubernetes.io/zone" # If not included, all zones are considered
      operator: In
      values: ["us-west-2a", "us-west-2b"]
    - key: "kubernetes.io/arch" # If not included, reference cloud provider
      operator: In
      values: ["arm64", "amd64"]

>     - key: "kubernetes.io/os" # only linux is currently supported

Reviewer 2: OS needs to be removed entirely. You can no longer specify this.

      operator: In
      values: ["linux"]
    - key: "karpenter.sh/capacity-type" # If not included, reference the cloud provider
      operator: In
      values: ["spot", "on-demand"]
  # These fields vary per cloud provider, see your cloud provider specific documentation
  provider: {}
```

## spec.requirements

Kubernetes defines the following [well known labels]([[link]]), and cloud providers (e.g., AWS) implement them. They are defined at the "spec.requirements" section of the provisioner CRD. 

These well known labels may be specified at the provisioner level, or in a workload definition (e.g., nodeSelector on a pod.spec). If the two conflict, the provisioner controls. In other words, a podspec may not override the provisioner with respect to constraints. 

> For example, the instance type may be specified using a nodeSelector in a workload definition. If the instance type requested at the workload (e.g., pod) level is not included in the provisioner list, karpenter will not create a node or schedule the pod. 

Reviewer 3: ```suggestion
For example, the instance type may be specified using a nodeSelector in a workload definition. If the instance type requested at the workload (i.e., pod) level is not included in the provisioner list, Karpenter will not create a node or schedule the pod. 
```

📝 None of these values are required. 

### Instance Types

> - key: `topology.kubernetes.io/instance-type`

Reviewer 2: ```suggestion
- key: `node.kubernetes.io/instance-type`
```

Generally, instance types should be a list and not a single value. Leaving this field undefined is recommended, as it maximizes choices for efficiently placing pods. 

☁️ **AWS**

Review [AWS instance types](https://aws.amazon.com/ec2/instance-types/).

The default value includes all instance types with the exclusion of metal
(non-virtualized),
[non-HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html),
and GPU instances.

View the full list of instance types with `aws ec2 describe-instance-types`.

**Example**

*Set Default with provisioner.yaml*

```yaml
spec:
  requirements:
    - key: node.kubernetes.io/instance-type
      operator: In
      values: ["m5.large", "m5.2xlarge"]
```

*Override with workload manifest (e.g., pod)*

```yaml
spec:
  template:
    spec:
      nodeSelector:
        node.kubernetes.io/instance-type: m5.large
```

### Availability Zones

- key: `topology.kubernetes.io/zone`
- value example: `us-east-1c`

☁️ **AWS**

- value list: `aws ec2 describe-availability-zones --region <region-name>`

> Karpenter can be configured to create nodes in a particular zone. Note that the Availability Zone us-east-1a for your AWS account might not have the same location as us-east-1a for another AWS account.

Reviewer 3: ```suggestion
Karpenter can be configured to create nodes in a particular zone. Note that the Availability Zone `us-east-1a` for your AWS account might not have the same location as `us-east-1a` for another AWS account.
```
I'm not sure if I've been consistent, but I generally like to mark literal values, such as `us-east-1a`, with backticks. I leave that to you.

[Learn more about Availability Zone
IDs.](https://docs.aws.amazon.com/ram/latest/userguide/working-with-az-ids.html)

### Architecture

- key: `kubernetes.io/arch`
- values
  - `amd64` (default)
  - `arm64`

Karpenter supports `amd64` nodes, and `arm64` nodes.

> ### Operating System
> 
> - key: `kubernetes.io/os`
> - values
>   - `linux` (default)
> 
> At this time, Karpenter only supports Linux OS nodes.

Reviewer 2: ```suggestion
```

### Capacity Type

- key: `karpenter.sh/capacity-type`

☁️ **AWS**

- values
  - `spot` (default)
  - `on-demand` 

> Karpenter supports specifying capacity type, which is analogous to EC2 usage classes.

Reviewer 2: ```suggestion
Karpenter supports specifying capacity type, which is analogous to [EC2 purchase types](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-purchasing-options.html).
```

> 🔶 *Karpenter defaults to spot instances.* [Spot instances](https://aws.amazon.com/ec2/spot/) may be preempted, and should not be used for critical workloads that do not tolerate interruptions.

Reviewer 2: This is inaccurate. The AWS provider defaults to OnDemand https://github.com/aws/karpenter/blob/main/pkg/cloudprovider/aws/apis/v1alpha1/provider_defaults.go#L47
